R/cpp_opts.R

```diff
@@ -0,0 +1,197 @@
+# run <model> info command
+run_info_cli <- function(exe_file) {
+  withr::with_path(
+    c(
+      toolchain_PATH_env_var(),
+      tbb_path()
+    ),
+    ret <- wsl_compatible_run(
+      command = wsl_safe_path(exe_file),
+      args = "info",
+      echo = is_verbose_mode(),
+      error_on_status = FALSE
+    )
+  )
+  ret
+}
+
+# new (future) parser
+# Parse the string output of <model> `info` into an R object (list)
+parse_exe_info_string <- function(ret_stdout) {
+  info <- list()
+  info_raw <- strsplit(strsplit(ret_stdout, "\n")[[1]], "=")
+  for (key_val in info_raw) {
+    if (length(key_val) > 1) {
+      key_val <- trimws(key_val)
+      val <- key_val[2]
+      if (!is.na(as.logical(val))) {
+        val <- as.logical(val)
+      }
+      info[[tolower(key_val[1])]] <- val
+    }
+  }
+
+  info[["stan_version"]] <- paste0(
+    info[["stan_version_major"]],
+    ".",
+    info[["stan_version_minor"]],
+    ".", info[["stan_version_patch"]]
+  )
+  info[["stan_version_major"]] <- NULL
+  info[["stan_version_minor"]] <- NULL
+  info[["stan_version_patch"]] <- NULL
+
+  info
+}
+
+# old (current) parser
+model_compile_info <- function(exe_file, version) {
+  info <- NULL
+  if (version > "2.26.1") {
+    
+    ret <- run_info_cli(exe_file)
+    if (ret$status == 0) {
+      info <- list()
+      info_raw <- strsplit(strsplit(ret$stdout, "\n")[[1]], "=")
+      for (key_val in info_raw) {
+        if (length(key_val) > 1) {
+          key_val <- trimws(key_val)
+          val <- key_val[2]
+          if (!is.na(as.logical(val))) {
+            val <- as.logical(val)
+          }
+          info[[toupper(key_val[1])]] <- val
+        }
+      }
+      info[["STAN_VERSION"]] <- paste0(info[["STAN_VERSION_MAJOR"]], ".", info[["STAN_VERSION_MINOR"]], ".", info[["STAN_VERSION_PATCH"]])
+      info[["STAN_VERSION_MAJOR"]] <- NULL
+      info[["STAN_VERSION_MINOR"]] <- NULL
+      info[["STAN_VERSION_PATCH"]] <- NULL
+    }
+  }
+  info
+}
+
+# convert to compile flags --------------------
+cpp_options_to_compile_flags <- function(cpp_options) {
+  if (length(cpp_options) == 0) {
+    return(NULL)
+  }
+  cpp_built_options <- c()
+  for (i in seq_along(cpp_options)) {
+    option_name <- names(cpp_options)[i]
+    if (is.null(option_name) || !nzchar(option_name)) {
+      cpp_built_options <- c(cpp_built_options, cpp_options[[i]])
+    } else {
+      cpp_built_options <- c(cpp_built_options, paste0(toupper(option_name), "=", cpp_options[[i]]))
+    }
+  }
+  cpp_built_options
+}
+
+
+# check options overall for validity ---------------------------------
+
+validate_cpp_options <- function(cpp_options) {
+  if (is.null(cpp_options) || length(cpp_options) == 0) return(list())
+
+  if (
+    !is.null(cpp_options[["user_header"]]) &&
+      !is.null(cpp_options[["USER_HEADER"]])
+  ) {
+    warning(
+      "User header specified both via cpp_options[[\"USER_HEADER\"]] ",
+      "and cpp_options[[\"user_header\"]].",
```

Comment:
When the user has both set do we include both? Or do we just take the distinct between them?
```suggestion
      "User header specified both via cpp_options[[\"USER_HEADER\"]] ",
      "and cpp_options[[\"user_header\"]]. Please only specify your user header in one location",
```